### DISCUSSION TITLE
Lesson 13 (proxies):  Why BoxV1 & BoxV2's constructor/initializer do not have arguments, but yet, when we refer to the proxy, we do things like -> BoxV2(proxy) or BoxV1(proxy)

### DISCUSSION THREAD

**Discussion by Glow-in-the-dark** [Cyfrin/foundry-full-course-cu#543] [Q&A] [answered]
![Screenshot 2023-08-04 at 11 39 58 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/2751458/2f31a593-ff10-4a9d-984c-5ed5aa7996b4)

Why when we are referring to the implementation, we do` BoxV2(proxy).getNumber()`, and not something like:
`proxy.getNumber()` ( making call to proxy since proxy is the main contract that owns the implementation, and storage is actually with the proxy contract )
or
'BoxV2.getNumber()' ( since the .getNumber function is at the implementation contract )

I suppose when we do things like BoxV1(proxy) or BoxV2(proxy). we are actually referring to the Proxy contract ? but it is wrapped up in the BoxV1 and BoxV2 format ? so we can more easily access the ABI ?

is there a more correct or intutive way to think about it ?

**Comment by theirrationalone** [comment, marked as answer]
Yes.... 

You can think proxy as the frontEnd.... 
BoxV1 and BoxV2 as two different backends where i like to define them as...

BoxV1 is downgraded due to poor APIs support. 
BoxV2 is an updated backend with rich & faster APIs support..... 

Now, as developers, we want to make an api call to updated backend... And we have both backend's api endpoints... 

So it's our duty to implement updated backend's api endpoints where ever we want to fetch as well as send data.

For example: backend that's now outdated had an api endpoint by using that we can(devs) let users(final consumers) to login. But the problem is, the implementation of login method has poor validation & safety support. And this backend has many other problems & bugs Or might missing some features yet to be implemented due to heavy demand. 

So we as the backend/fullstack developers team decided to develop a new backend with heavy support and features. Now we have two backends, we will update all endpoints with new ones. 

That's why (you may able to imagine now) we wrapped proxy with BoxV2, Although proxy & BoxV1 also have some same properties & method. 

# **Chillis:**

1. For developers usage we also have to update our documentations, and have to send a 🔔**NOTIFICATION** to all developers as consumers.Then so this method would be called as the informing method, Hey everybody we have updated our contract now it's not BoxV1 instead is BoxV2.

2. For end users (Final consumers), they don't have any idea about that update and they don't need to know about. They just use the services, Proxy as an abstract. So this method would be called as The UUPs.

**Reply by Glow-in-the-dark** [reply]
mmmm what do u mean by notification? could u expand on that ?

**Reply by theirrationalone** [reply]
By notification i meant...: For Example, You have subscribed ReactJs's newsletter, and there is a new critical update about ReacJs's. or let's say they released new version of ReactJs.

Here at that point they won't need to notify all the Web2's or may WEB3's sites(can't say frontend only because ReactJs also has its framework for working with backend(NextJs)) **OWNER** (The Person or Company or Org etc, Who owns **FullStack**. Concerned only with his/her/its Business(Abstract), Owner might hired devs to develop Business Engine.) and all the site's consumer who're using those websites, They have only One **DUTY / RESPONSIBILITY** to inform all the **WEB-DEVs** through their **NEWS MEDIUMS**.

You as a ReactJs Dev😎, In your inbox....

# **ReactJs: Hey Devs, It's fun time We just released our Latest version of ReactJs... Go Check it out on our official website!**
